File: src/fogbugz-client.ts
import type {
  FogBugzResponse,
  SearchData,
  CaseActionData,
  ListPeopleData,
  ViewPersonData,
  ListProjectsData,
  ListFixForsData,
  ListKanbanColumnsData,
  ListCategoriesData,
  ListStatusesData,
  ListPrioritiesData,
} from "./types.js";

export class FogBugzApiError extends Error {
  constructor(
    public code: string,
    message: string,
  ) {
    super(message);
    this.name = "FogBugzApiError";
  }
}

export class FogBugzClient {
  private baseUrl: string;
  private token: string;

  constructor(baseUrl: string, token: string) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
    this.token = token;
  }

  private async request<T>(
    command: string,
    params: Record<string, unknown> = {},
  ): Promise<T> {
    const url = `${this.baseUrl}/api/${command}`;
    const body = { token: this.token, ...params };

    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json; charset=utf-8" },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      throw new FogBugzApiError(
        "HTTP_ERROR",
        `HTTP ${response.status}: ${response.statusText}`,
      );
    }

    const json = (await response.json()) as FogBugzResponse<T>;

    if (json.errors && json.errors.length > 0) {
      const err = json.errors[0];
      throw new FogBugzApiError(err.code, err.message);
    }

    return json.data;
  }

  async search(
    q: string,
    cols: string[],
    max?: number,
  ): Promise<SearchData> {
    return this.request<SearchData>("search", {
      q,
      cols,
      ...(max !== undefined ? { max } : {}),
    });
  }

  async edit(
    ixBug: number,
    fields: Record<string, unknown>,
  ): Promise<CaseActionData> {
    return this.request<CaseActionData>("edit", { ixBug, ...fields });
  }

  async assign(
    ixBug: number,
    assignee: string,
    sEvent?: string,
  ): Promise<CaseActionData> {
    return this.request<CaseActionData>("assign", {
      ixBug,
      sPersonAssignedTo: assignee,
      ...(sEvent ? { sEvent } : {}),
    });
  }

  async resolve(
    ixBug: number,
    options?: { ixStatus?: number; sStatus?: string; sEvent?: string },
  ): Promise<CaseActionData> {
    return this.request<CaseActionData>("resolve", {
      ixBug,
      ...options,
    });
  }

  async close(
    ixBug: number,
    sEvent?: string,
  ): Promise<CaseActionData> {
    return this.request<CaseActionData>("close", {
      ixBug,
      ...(sEvent ? { sEvent } : {}),
    });
  }

  async reopen(
    ixBug: number,
    sEvent?: string,
  ): Promise<CaseActionData> {
    return this.request<CaseActionData>("reopen", {
      ixBug,
      ...(sEvent ? { sEvent } : {}),
    });
  }

  async reactivate(
    ixBug: number,
    sEvent?: string,
  ): Promise<CaseActionData> {
    return this.request<CaseActionData>("reactivate", {
      ixBug,
      ...(sEvent ? { sEvent } : {}),
    });
  }

  async listPeople(options?: {
    fIncludeActive?: number;
    fIncludeNormal?: number;
    fIncludeDeleted?: number;
    fIncludeCommunity?: number;
    fIncludeVirtual?: number;
  }): Promise<ListPeopleData> {
    return this.request<ListPeopleData>("listPeople", options ?? {});
  }

  async viewPerson(lookup?: {
    sEmail?: string;
    sFullname?: string;
  }): Promise<ViewPersonData> {
    return this.request<ViewPersonData>("viewPerson", lookup ?? {});
  }

  async listProjects(): Promise<ListProjectsData> {
    return this.request<ListProjectsData>("listProjects");
  }

  async listFixFors(ixProject?: number): Promise<ListFixForsData> {
    return this.request<ListFixForsData>("listFixFors", {
      ...(ixProject !== undefined ? { ixProject } : {}),
    });
  }

  async listKanbanColumns(
    ixPlanner?: number,
  ): Promise<ListKanbanColumnsData> {
    return this.request<ListKanbanColumnsData>("listKanbanColumns", {
      ...(ixPlanner !== undefined ? { ixPlanner } : {}),
    });
  }

  async listCategories(): Promise<ListCategoriesData> {
    return this.request<ListCategoriesData>("listCategories");
  }

  async listStatuses(): Promise<ListStatusesData> {
    return this.request<ListStatusesData>("listStatuses");
  }

  async listPriorities(): Promise<ListPrioritiesData> {
    return this.request<ListPrioritiesData>("listPriorities");
  }

  getCaseUrl(ixBug: number): string {
    return `${this.baseUrl}/f/cases/${ixBug}`;
  }
}

File: python/src/fogbugz_mcp/fogbugz_client.py
from __future__ import annotations

from typing import Any

import httpx

from .types import CaseActionData, ListPeopleData, SearchData


class FogBugzApiError(Exception):
    def __init__(self, code: str, message: str) -> None:
        super().__init__(message)
        self.code = code


class FogBugzClient:
    def __init__(self, base_url: str, token: str) -> None:
        self._base_url = base_url.rstrip("/")
        self._token = token
        self._http = httpx.AsyncClient()

    async def _request(self, command: str, params: dict[str, Any] | None = None) -> Any:
        url = f"{self._base_url}/api/{command}"
        body: dict[str, Any] = {"token": self._token}
        if params:
            body.update(params)

        response = await self._http.post(
            url,
            json=body,
            headers={"Content-Type": "application/json; charset=utf-8"},
        )

        if not response.is_success:
            raise FogBugzApiError(
                "HTTP_ERROR",
                f"HTTP {response.status_code}: {response.reason_phrase}",
            )

        data = response.json()
        errors = data.get("errors") or []
        if errors:
            err = errors[0]
            raise FogBugzApiError(err["code"], err["message"])

        return data["data"]

    async def search(self, q: str, cols: list[str], max: int | None = None) -> SearchData:
        params: dict[str, Any] = {"q": q, "cols": cols}
        if max is not None:
            params["max"] = max
        return await self._request("search", params)

    async def new_case(self, fields: dict[str, Any]) -> CaseActionData:
        return await self._request("new", fields)

    async def edit(self, ix_bug: int, fields: dict[str, Any]) -> CaseActionData:
        return await self._request("edit", {"ixBug": ix_bug, **fields})

    async def assign(self, ix_bug: int, assignee: str, s_event: str | None = None) -> CaseActionData:
        params: dict[str, Any] = {"ixBug": ix_bug, "sPersonAssignedTo": assignee}
        if s_event:
            params["sEvent"] = s_event
        return await self._request("assign", params)

    async def resolve(self, ix_bug: int, options: dict[str, Any] | None = None) -> CaseActionData:
        params: dict[str, Any] = {"ixBug": ix_bug}
        if options:
            params.update(options)
        return await self._request("resolve", params)

    async def close(self, ix_bug: int, s_event: str | None = None) -> CaseActionData:
        params: dict[str, Any] = {"ixBug": ix_bug}
        if s_event:
            params["sEvent"] = s_event
        return await self._request("close", params)

    async def reopen(self, ix_bug: int, s_event: str | None = None) -> CaseActionData:
        params: dict[str, Any] = {"ixBug": ix_bug}
        if s_event:
            params["sEvent"] = s_event
        return await self._request("reopen", params)

    async def reactivate(self, ix_bug: int, s_event: str | None = None) -> CaseActionData:
        params: dict[str, Any] = {"ixBug": ix_bug}
        if s_event:
            params["sEvent"] = s_event
        return await self._request("reactivate", params)

    async def list_people(self, options: dict[str, Any] | None = None) -> ListPeopleData:
        return await self._request("listPeople", options or {})

    def get_case_url(self, ix_bug: int) -> str:
        return f"{self._base_url}/f/cases/{ix_bug}"
